lib/plugins/aws/package/compile/events/apiGateway/lib/requestValidator.js

```diff
@@ -0,0 +1,216 @@
+            modelLogicalId
+          ].Properties.Description = description;
+        }
+      }
+    }
+    this.validated.events.forEach(event => {
+      const resourceName = this.getResourceName(event.http.path);
+      const methodLogicalId = this.provider.naming.getMethodLogicalId(
+        resourceName,
+        event.http.method
+      );
+
+      // Old functionality
+      if (event.http.request && event.http.request.schema) {
+        const template = this.serverless.service.provider.compiledCloudFormationTemplate.Resources[
+          methodLogicalId
+        ];
+        for (const [contentType, schema] of _.entries(event.http.request.schema)) {
+          const modelLogicalId = this.provider.naming.getModelLogicalId(
+            resourceName,
+            event.http.method,
+            contentType
+          );
+
+          const validatorLogicalId = this.provider.naming.getValidatorLogicalId(
+            resourceName,
+            event.http.method
+          );
+
+          template.Properties.RequestValidatorId = { Ref: validatorLogicalId };
+          template.Properties.RequestModels = template.Properties.RequestModels || {};
+          template.Properties.RequestModels[contentType] = { Ref: modelLogicalId };
+
+          _.merge(this.serverless.service.provider.compiledCloudFormationTemplate.Resources, {
+            [modelLogicalId]: {
+              Type: 'AWS::ApiGateway::Model',
+              Properties: {
+                RestApiId: this.provider.getApiGatewayRestApiId(),
+                ContentType: contentType,
+                Schema: schema,
+              },
+            },
+            [validatorLogicalId]: {
+              Type: 'AWS::ApiGateway::RequestValidator',
+              Properties: {
+                RestApiId: this.provider.getApiGatewayRestApiId(),
+                ValidateRequestBody: true,
+                ValidateRequestParameters: true,
+              },
+            },
+          });
+        }
+      }
+
+      // New Functionality
+      if (event.http.requestSchema) {
+        const template = this.serverless.service.provider.compiledCloudFormationTemplate.Resources[
+          methodLogicalId
+        ];
+        for (const [contentType, schemaConfig] of _.entries(event.http.requestSchema)) {
+          let modelName;
+          let description;
+          let definition;
+          let validatorName;
+
+          if (schemaConfig.schema) {
+            modelName = schemaConfig.name ? schemaConfig.name : null;
+            validatorName = modelName !== null ? `${modelName}Validator` : null;
+            description = schemaConfig.description ? schemaConfig.description : null;
+            definition = schemaConfig.schema;
+          } else {
+            definition = schemaConfig;
+          }
+
+          let modelLogicalId;
+          let validatorLogicalId;
+
+          // New resources need to be created
+          if (!models[schemaConfig]) {
```

Comment:
Logic seems to be a bit leaky, as we're also running this check if `schemaConfig` is object.

I propose to restructure above logic into:

```javascript
if (_.isObject(schemaConfig)) {
  // Schema defined inline, specific to given endpont
} else {
  // String
  if (models[schemaConfig]) {
    // Schema reused
  }  else {
    // Simple schema defined inline, specific to given endpoint
  }
}
```

Reply:
I will add an `or` to it as `if (_.isObject(schemaConfig) || !models[schemaConfig])`. I was trying to DRY out as much as possible without writing a bunch of helper methods.